In [ ]:
import torch
print(torch.cuda.is_available())
print(torch.cuda.get_device_name(0) if torch.cuda.is_available() else "No GPU")

True
Tesla T4


In [ ]:
import os

DATA_DIR = "/content/work"
RUNS_DIR = os.path.join(DATA_DIR, "step5_runs")
RUN_DIR = os.path.join(RUNS_DIR, "coco_to_cityscapes_semantic_ft")
CKPT_DIR = os.path.join(RUN_DIR, "checkpoints")

os.makedirs(DATA_DIR, exist_ok=True)
os.makedirs(RUNS_DIR, exist_ok=True)
os.makedirs(RUN_DIR, exist_ok=True)
os.makedirs(CKPT_DIR, exist_ok=True)

CITY_CKPT = os.path.join(DATA_DIR, "eomt_cityscapes.bin")
COCO_CKPT = os.path.join(DATA_DIR, "eomt_coco.bin")

print("DATA_DIR:", DATA_DIR)
print("RUN_DIR:", RUN_DIR)
print("CKPT_DIR:", CKPT_DIR)
print("CITY_CKPT exists:", os.path.exists(CITY_CKPT))
print("COCO_CKPT exists:", os.path.exists(COCO_CKPT))

DATA_DIR: /content/work
RUN_DIR: /content/work/step5_runs/coco_to_cityscapes_semantic_ft
CKPT_DIR: /content/work/step5_runs/coco_to_cityscapes_semantic_ft/checkpoints
CITY_CKPT exists: False
COCO_CKPT exists: False


In [ ]:
import getpass

CITY_USER = input("Cityscapes username: ")
CITY_PASS = getpass.getpass("Cityscapes password: ")

Cityscapes username: belfo
Cityscapes password: ··········


In [ ]:
!wget --keep-session-cookies \
  --save-cookies cookies.txt \
  --post-data "username=$CITY_USER&password=$CITY_PASS&submit=Login" \
  "https://www.cityscapes-dataset.com/login/" \
  -O login_response.html

!wget -c --load-cookies cookies.txt --content-disposition \
  "https://www.cityscapes-dataset.com/file-handling/?packageID=1" \
  -O /content/work/gtFine_trainvaltest.zip

!wget -c --load-cookies cookies.txt --content-disposition \
  "https://www.cityscapes-dataset.com/file-handling/?packageID=3" \
  -O /content/work/leftImg8bit_trainvaltest.zip

--2026-04-23 08:50:35--  https://www.cityscapes-dataset.com/login/
Resolving www.cityscapes-dataset.com (www.cityscapes-dataset.com)... 139.19.217.8
Connecting to www.cityscapes-dataset.com (www.cityscapes-dataset.com)|139.19.217.8|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.cityscapes-dataset.com/downloads/ [following]
--2026-04-23 08:50:35--  https://www.cityscapes-dataset.com/downloads/
Reusing existing connection to www.cityscapes-dataset.com:443.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘login_response.html’

login_response.html     [ <=>                ]  55.82K  --.-KB/s    in 0.06s   

2026-04-23 08:50:35 (903 KB/s) - ‘login_response.html’ saved [57162]

--2026-04-23 08:50:35--  https://www.cityscapes-dataset.com/file-handling/?packageID=1
Resolving www.cityscapes-dataset.com (www.cityscapes-dataset.com)... 139.19.217.8
Connecting to www.cityscapes-dataset.com (www.cityscapes-dataset

In [ ]:
!pip install -q gdown
!gdown --fuzzy "https://drive.google.com/file/d/1X57RfPr5FCemAIYAeNeev3AWPaW5Pjb1/view?usp=sharing" -O /content/work/eomt_cityscapes.bin
!gdown --fuzzy "https://drive.google.com/file/d/1fF-bevMQjDjCb15xDoYQGPpVqJZNY0XO/view?usp=sharing" -O /content/work/eomt_coco.bin

Downloading...
From (original): https://drive.google.com/uc?id=1X57RfPr5FCemAIYAeNeev3AWPaW5Pjb1
From (redirected): https://drive.google.com/uc?id=1X57RfPr5FCemAIYAeNeev3AWPaW5Pjb1&confirm=t&uuid=45a9a1f9-fcf7-46c7-b976-55f3708e140c
To: /content/work/eomt_cityscapes.bin
100% 382M/382M [00:04<00:00, 79.6MB/s]
Downloading...
From (original): https://drive.google.com/uc?id=1fF-bevMQjDjCb15xDoYQGPpVqJZNY0XO
From (redirected): https://drive.google.com/uc?id=1fF-bevMQjDjCb15xDoYQGPpVqJZNY0XO&confirm=t&uuid=c0a015cb-0efa-4f9d-ab05-8d6b767eef4d
To: /content/work/eomt_coco.bin
100% 375M/375M [00:04<00:00, 93.3MB/s]


In [ ]:
!git clone https://github.com/AlessandroMarinai/MaskArchitectureAnomaly_CourseProject.git /content/project

Cloning into '/content/project'...
remote: Enumerating objects: 131, done.
remote: Counting objects: 100% (45/45), done.
remote: Compressing objects: 100% (38/38), done.
remote: Total 131 (delta 14), reused 7 (delta 7), pack-reused 86 (from 4)
Receiving objects: 100% (131/131), 26.88 MiB | 29.98 MiB/s, done.
Resolving deltas: 100% (23/23), done.


In [ ]:
%cd /content/project/eomt
!pip install -r requirements.txt

/content/project/eomt
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.0/62.0 kB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 219.4/219.4 kB 18.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.6/8.6 MB 87.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 37.3/37.3 MB 32.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 865.0/865.0 MB 1.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.4/7.4 MB 81.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.2/117.2 kB 12.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 393.1/393.1 MB 1.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 76.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 59.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 897.7/89

# Baseline validation of the original Cityscapes model

On the 500 images of the Cityscapes validation set, the model obtains a mean Intersection over Union of approximately 81.7% on the semantic segmentation classes


In [ ]:
# Baseline validation of the original Cityscapes model on full Cityscapes classes
%cd /content/project/eomt

!python3 main.py validate \
  -c configs/dinov2/cityscapes/semantic/eomt_base_640.yaml \
  --trainer.devices 1 \
  --data.batch_size 1 \
  --data.path "/content/work" \
  --model.ckpt_path "/content/work/eomt_cityscapes.bin"

/content/project/eomt
2026-04-23 09:00:42.991781: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1776934843.253035    4286 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1776934843.320109    4286 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1776934843.860128    4286 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776934843.860178    4286 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776934843.860192    4286 computation_placer.cc:177] 

# Fine-tuning the COCO model on Cityscapes semantic segmentation


In [ ]:
import copy
import yaml
import pprint
import os

CITY_CFG_PATH = "/content/project/eomt/configs/dinov2/cityscapes/semantic/eomt_base_640.yaml"
COCO_CFG_PATH = "/content/project/eomt/configs/dinov2/coco/panoptic/eomt_base_640_2x.yaml"
CUSTOM_CFG_PATH = "/content/project/eomt/configs/dinov2/cityscapes/semantic/eomt_base_640_from_coco_ft.yaml"

RUN_DIR = "/content/work/step5_runs/coco_to_cityscapes_semantic_ft"
CKPT_DIR = os.path.join(RUN_DIR, "checkpoints")
WANDB_DIR = os.path.join(RUN_DIR, "wandb")

os.makedirs(RUN_DIR, exist_ok=True)
os.makedirs(CKPT_DIR, exist_ok=True)
os.makedirs(WANDB_DIR, exist_ok=True)

with open(CITY_CFG_PATH, "r") as f:
    city_cfg = yaml.safe_load(f)

with open(COCO_CFG_PATH, "r") as f:
    coco_cfg = yaml.safe_load(f)

ft_cfg = copy.deepcopy(city_cfg)

# Replace only the internal network definition with the COCO-compatible one to keep the task
# as Cityscapes semantic segmentation while allowing us to initialize from the COCO-trained checkpoint
ft_cfg["model"]["init_args"]["network"] = copy.deepcopy(
    coco_cfg["model"]["init_args"]["network"]
)

# Check if the data section already contains an "init_args" dictionary
if "init_args" not in ft_cfg["data"]:
    ft_cfg["data"]["init_args"] = {}

# Optional: use 2 dataloader workers to reduce Colab
ft_cfg["data"]["init_args"]["num_workers"] = 2

# Replace the trainer section with a clean fine-tuning setup.
ft_cfg["trainer"] = {
    "default_root_dir": RUN_DIR,        # where logs/checkpoints are saved
    "devices": 1,                       # use one GPU
    "precision": "16-mixed",            # AMP / mixed precision
    "max_epochs": 3,                    # number of epochs
    "accumulate_grad_batches": 4,       # gradient accumulation
    "enable_checkpointing": True,       # save checkpoints

    # Configure the experiment logger to keep trace of the metrics
    "logger": {
        "class_path": "lightning.pytorch.loggers.WandbLogger",  #wandb.wandbLoggers (nello yaml su github)
        "init_args": {
            "project": "step5-coco-to-cityscapes-ft",
            "name": "offline_ft_run",
            "save_dir": WANDB_DIR,
            "offline": True
        }
    },

    # to save the checkpoints
    "callbacks": [
        {
            "class_path": "lightning.pytorch.callbacks.ModelCheckpoint",
            "init_args": {
                "dirpath": CKPT_DIR,                # where checkpoints are saved
                "filename": "best",                 # best checkpoint name
                "monitor": "metrics/val_iou_all",   # which metric to observe
                "mode": "max",                      # to look for the best (max) metric
                "save_top_k": 1,                    # to save only the best ckpt
                "save_last": True,                  # to save also the last ckpt
                "auto_insert_metric_name": False    # Evita che Lightning aggiunga automaticamente il nome della metrica nel filename?
            }
        }
    ]
}

with open(CUSTOM_CFG_PATH, "w") as f:
    yaml.safe_dump(ft_cfg, f, sort_keys=False)

print("Saved custom fine-tuning config to:")
print(CUSTOM_CFG_PATH)

print("\nData section:")
pprint.pprint(ft_cfg["data"])

print("\nTrainer section:")
pprint.pprint(ft_cfg["trainer"])

Saved custom fine-tuning config to:
/content/project/eomt/configs/dinov2/cityscapes/semantic/eomt_base_640_from_coco_ft.yaml

Data section:
{'class_path': 'datasets.cityscapes_semantic.CityscapesSemantic',
 'init_args': {'num_workers': 2}}

Trainer section:
{'accumulate_grad_batches': 4,
 'callbacks': [{'class_path': 'lightning.pytorch.callbacks.ModelCheckpoint',
                'init_args': {'auto_insert_metric_name': False,
                              'dirpath': '/content/work/step5_runs/coco_to_cityscapes_semantic_ft/checkpoints',
                              'filename': 'best',
                              'mode': 'max',
                              'monitor': 'metrics/val_iou_all',
                              'save_last': True,
                              'save_top_k': 1}}],
 'default_root_dir': '/content/work/step5_runs/coco_to_cityscapes_semantic_ft',
 'devices': 1,
 'enable_checkpointing': True,
 'logger': {'class_path': 'lightning.pytorch.loggers.WandbLogger',
       

# First training run of the COCO model fine-tuned on Cityscapes semantic segmentation


Epoch 0: train_loss_total=12.40, mIoU: 55.5

Epoch 1: train_loss_total=10.50, mIoU: 70.2

Epoch 2: ...

In [ ]:
%cd /content/project/eomt

# Run fine-tuning from the COCO checkpoint using the custom Cityscapes semantic configuration
!WANDB_MODE=offline python3 main.py fit \
  -c /content/project/eomt/configs/dinov2/cityscapes/semantic/eomt_base_640_from_coco_ft.yaml \
  --data.path "/content/work" \
  --data.batch_size 1 \
  --data.img_size [640,640] \
  --model.ckpt_path "/content/work/eomt_coco.bin" \
  --model.load_ckpt_class_head False

/content/project/eomt
2026-04-23 09:23:08.795354: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1776936188.817831    9912 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1776936188.824414    9912 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1776936188.841000    9912 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776936188.841049    9912 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776936188.841055    9912 computation_placer.cc:177] 

In [ ]:
import os

best_ckpt = "/content/work/step5_runs/coco_to_cityscapes_semantic_ft/checkpoints/best.ckpt"
last_ckpt = "/content/work/step5_runs/coco_to_cityscapes_semantic_ft/checkpoints/last.ckpt"

FT_CKPT = best_ckpt if os.path.exists(best_ckpt) else last_ckpt
print("Using checkpoint:", FT_CKPT)

%cd /content/project/eomt

# Additional validation of the fine-tuned model on Cityscapes semantic segmentation
!python3 main.py validate \
  -c /content/project/eomt/configs/dinov2/cityscapes/semantic/eomt_base_640_from_coco_ft.yaml \
  --trainer.devices 1 \
  --data.batch_size 1 \
  --data.img_size [640,640] \
  --data.path "/content/work" \
  --model.ckpt_path "{FT_CKPT}"

Using checkpoint: /content/work/step5_runs/coco_to_cityscapes_semantic_ft/checkpoints/best.ckpt
/content/project/eomt
2026-04-23 10:40:10.675622: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1776940810.698166   30491 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1776940810.705223   30491 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1776940810.722856   30491 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1776940810.722906   30491 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking t